src/printer/let_binding.cpp

```diff
@@ -0,0 +1,216 @@
+
+uint32_t LetBinding::getId(Node n) const
+{
+  NodeIdMap::const_iterator it = d_letMap.find(n);
+  if (it == d_letMap.end())
+  {
+    return 0;
+  }
+  return (*it).second;
+}
+
+Node LetBinding::convert(Node n, const std::string& prefix, bool letTop) const
+{
+  if (d_letMap.empty())
+  {
+    return n;
+  }
+  NodeManager* nm = NodeManager::currentNM();
+  std::unordered_map<TNode, Node, TNodeHashFunction> visited;
+  std::unordered_map<TNode, Node, TNodeHashFunction>::iterator it;
+  std::vector<TNode> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    visit.pop_back();
+    it = visited.find(cur);
+
+    if (it == visited.end())
+    {
+      uint32_t id = getId(cur);
+      // do not letify id 0, or n itself if letTop is false
+      if (id > 0 && (cur != n || letTop))
+      {
+        // make the let variable
+        std::stringstream ss;
+        ss << prefix << id;
+        visited[cur] = nm->mkBoundVar(ss.str(), cur.getType());
+      }
+      else
+      {
+        visited[cur] = Node::null();
+        visit.push_back(cur);
+        visit.insert(visit.end(), cur.begin(), cur.end());
+      }
+    }
+    else if (it->second.isNull())
+    {
+      Node ret = cur;
+      bool childChanged = false;
+      std::vector<Node> children;
+      if (cur.getMetaKind() == kind::metakind::PARAMETERIZED)
+      {
+        children.push_back(cur.getOperator());
+      }
+      for (const Node& cn : cur)
+      {
+        it = visited.find(cn);
+        Assert(it != visited.end());
+        Assert(!it->second.isNull());
+        childChanged = childChanged || cn != it->second;
+        children.push_back(it->second);
+      }
+      if (childChanged)
+      {
+        ret = nm->mkNode(cur.getKind(), children);
+      }
+      visited[cur] = ret;
+    }
+  } while (!visit.empty());
+  Assert(visited.find(n) != visited.end());
+  Assert(!visited.find(n)->second.isNull());
+  return visited[n];
+}
+
+void LetBinding::updateCounts(Node n)
+{
+  NodeIdMap::iterator it;
+  std::vector<Node> visit;
+  TNode cur;
+  visit.push_back(n);
+  do
+  {
+    cur = visit.back();
+    it = d_count.find(cur);
+    if (it == d_count.end())
+    {
+      // do not traverse beneath quantifiers
+      if (cur.getNumChildren() == 0 || cur.isClosure())
+      {
+        d_visitList.push_back(cur);
+        d_count[cur] = 1;
+        visit.pop_back();
+      }
+      else
+      {
+        d_count[cur] = 0;
+        visit.insert(visit.end(), cur.begin(), cur.end());
+      }
+    }
+    else
+    {
+      if ((*it).second == 0)
+      {
+        d_visitList.push_back(cur);
+      }
+      d_count[cur] = (*it).second + 1;
+      visit.pop_back();
+    }
+  } while (!visit.empty());
+}
+
+void LetBinding::convertCountToLet()
+{
+  Assert(d_thresh > 0);
+  // Assign ids for those whose d_count is >= d_thresh, traverse in d_visitList
+  // in order so that deeper proofs are assigned lower identifiers, which
```

Comment:
deeper proofs?

Reply:
Will fix, this was copied from the proof letifier accidentally.